### PR TITLE
HHH-17322 Allow custom generators to determine timing based on current entity state

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-17322

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/generator/Generator.java [hunk 1]
```diff
@@ -93,6 +96,32 @@ public interface Generator extends Serializable {
...
+	 * @return {@code true} if the value is generated by the database as a side effect of
+	 * the execution of an {@code insert} or {@code update} statement, or false if
+	 * it is generated in Java code before the statement is executed via JDBC.
+	 *
+	 * @see #generatedOnExecution()
+	 * @see BeforeExecutionGenerator
+	 * @see OnExecutionGenerator
+	 * @since 6.4
+	 */
+	default boolean generatedOnExecution(SharedSessionContractImplementor session, Object entity) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/generator/Generator.java:121]: Usually, we pass context objects like `session` last, but that's a minor note.

**Review by beikov — APPROVED** [review]

File: hibernate-core/src/main/java/org/hibernate/generator/Generator.java [hunk 2]
```diff
@@ -16,7 +16,7 @@
 
 /**
  * Describes the generation of values of a certain field or property of an entity. A generated
- * value might be generated in Java, or by the database. Every instance must implement either
+ * value might be generated in Java, or by the database. Every instance of this interface can implement
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/generator/Generator.java:19]: Perhaps:

> Every instance must implement either of both of

File: hibernate-core/src/main/java/org/hibernate/generator/Generator.java [hunk 3]
```diff
@@ -34,6 +34,9 @@
  *	   SQL {@code select}, though in certain cases this additional round trip may be avoided.
  *	   An important example is id generation using an identity column.
  * </ul>
+ * Note that it's also possible for a Generator to implement both interfaces and determine the timing
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/generator/Generator.java:37]: "A Generator may implement" is enough.

"Note that it's also possible for" is not really doing any useful work here.

**Comment by gavinking** [comment]
This change looks good!

File: hibernate-core/src/main/java/org/hibernate/generator/Generator.java (hunk 3 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/generator/Generator.java:37]: Thanks for the suggestion, applied it.

File: hibernate-core/src/main/java/org/hibernate/generator/Generator.java (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/generator/Generator.java]: Thanks for the suggestion, inverted the method parameters.

**Pull request merged**